types/src/client.rs

```diff
@@ -200,52 +201,70 @@ pub struct RequestIdGuard {
 impl RequestIdGuard {
 	/// Create a new `RequestIdGuard` with the provided concurrency limit.
 	pub fn new(limit: usize) -> Self {
-		Self { current_pending: AtomicUsize::new(0), max_concurrent_requests: limit, current_id: AtomicU64::new(0) }
+		Self { current_pending: Arc::new(()), max_concurrent_requests: limit, current_id: AtomicU64::new(0) }
 	}
 
-	fn get_slot(&self) -> Result<(), Error> {
-		self.current_pending
-			.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |val| {
-				if val >= self.max_concurrent_requests {
-					None
-				} else {
-					Some(val + 1)
-				}
-			})
-			.map(|_| ())
-			.map_err(|_| Error::MaxSlotsExceeded)
+	fn get_slot(&self) -> Result<Arc<()>, Error> {
+		// Strong count is 1 at start, so that's why we use `>` and not `>=`.
+		if Arc::strong_count(&self.current_pending) > self.max_concurrent_requests {
+			Err(Error::MaxSlotsExceeded)
+		} else {
+			Ok(self.current_pending.clone())
+		}
 	}
 
 	/// Attempts to get the next request ID.
 	///
 	/// Fails if request limit has been exceeded.
-	pub fn next_request_id(&self) -> Result<u64, Error> {
-		self.get_slot()?;
+	pub fn next_request_id(&self) -> Result<RequestId<u64>, Error> {
+		let rc = self.get_slot()?;
 		let id = self.current_id.fetch_add(1, Ordering::SeqCst);
-		Ok(id)
+		Ok(RequestId { _rc: rc, id })
 	}
 
 	/// Attempts to get the `n` number next IDs that only counts as one request.
 	///
 	/// Fails if request limit has been exceeded.
-	pub fn next_request_ids(&self, len: usize) -> Result<Vec<u64>, Error> {
-		self.get_slot()?;
-		let mut batch = Vec::with_capacity(len);
+	pub fn next_request_ids(&self, len: usize) -> Result<RequestId<Vec<u64>>, Error> {
+		let rc = self.get_slot()?;
+		let mut ids = Vec::with_capacity(len);
 		for _ in 0..len {
-			batch.push(self.current_id.fetch_add(1, Ordering::SeqCst));
+			ids.push(self.current_id.fetch_add(1, Ordering::SeqCst));
 		}
-		Ok(batch)
+		Ok(RequestId { _rc: rc, id: ids })
+	}
+}
+
+/// Reference counted request ID.
+#[derive(Debug)]
+pub struct RequestId<T> {
+	id: T,
+	/// Reference count decreased when dropped.
+	_rc: Arc<()>,
+}
+
+impl<T> RequestId<T> {
```

Comment:
I guess this could be a `Deref` impl too, but given how it's used I wouldn't mind either way!